### PR TITLE
feat: optimise BinningData, remove dead code

### PR DESCRIPTION
**Pull request by asalzburger**
This PR cleans the `BinUtility` and  `BinningData` objects:
- removes dead code
- removes hand-written search in vector in favour of `std::lower_bound`

It adds/removes relevant `UnitTests` and adds a new benchmark test, the benchmark before the changes shows:
```shell
16:36:17    BinUtility     INFO      Execution stats small: 1000 runs of 1 iteration(s), 0.0ms total, 0.0430+/-0.0001µs per run, 43.000+/-0.058ns per iteration
16:36:17    BinUtility     INFO      Fraction is: 21968011 vs. 21968011
16:36:19    BinUtility     INFO      Execution stats medium: 1000 runs of 1 iteration(s), 0.1ms total, 0.0510+/-0.0003µs per run, 51.000+/-0.346ns per iteration
16:36:19    BinUtility     INFO      Fraction is: 19641811 vs. 19641811
16:36:21    BinUtility     INFO      Execution stats many: 1000 runs of 1 iteration(s), 0.1ms total, 0.0780+/-0.0002µs per run, 78.000+/-0.173ns per iteration
16:36:21    BinUtility     INFO      Fraction is: 12780128 vs. 12780128
16:36:23    BinUtility     INFO      Execution stats many (lower_bounds): 1000 runs of 1 iteration(s), 0.0ms total, 0.0420+/-0.0001µs per run, 42.000+/-0.086ns per iteration
16:36:23    BinUtility     INFO      Fraction is: 23670634 vs. 23670634
```

to the new implementation:
```shell
17:25:25    BinUtility     INFO      Execution stats small: 1000 runs of 1 iteration(s), 0.0ms total, 0.0450+/-0.0001µs per run, 45.000+/-0.115ns per iteration
17:25:25    BinUtility     INFO      Fraction is: 21444179 vs. 21444179
17:25:27    BinUtility     INFO      Execution stats medium: 1000 runs of 1 iteration(s), 0.0ms total, 0.0460+/-0.0001µs per run, 46.000+/-0.115ns per iteration
17:25:27    BinUtility     INFO      Fraction is: 21030434 vs. 21030434
17:25:29    BinUtility     INFO      Execution stats many: 1000 runs of 1 iteration(s), 0.0ms total, 0.0480+/-0.0001µs per run, 48.000+/-0.115ns per iteration
17:25:29    BinUtility     INFO      Fraction is: 20028335 vs. 20028335
17:25:31    BinUtility     INFO      Execution stats equidistant: 1000 runs of 1 iteration(s), 0.0ms total, 0.0450+/-0.0001µs per run, 45.000+/-0.058ns per iteration
17:25:31    BinUtility     INFO      Fraction is: 20992746 vs. 20992746
```
The new implementation is faster in all bin scenarios.

Eventually, the `BinUtility` should go in favour of `Grid` usage, the removal of dead code is a first step towards this.

### CODE REVIEW

File: Tests/UnitTests/Core/Utilities/BinningDataTests.cpp [hunk 1]
```diff
@@ -197,7 +197,7 @@ BOOST_AUTO_TEST_CASE(BinningData_bins) {
   BOOST_CHECK_EQUAL(yData_eq.searchLocal(xyPosition), size_t(1));
   BOOST_CHECK_EQUAL(zData_eq.searchLocal(rphizPosition), size_t(2));
   BOOST_CHECK_EQUAL(xData_arb.searchLocal(xyPosition), size_t(0));
-  BOOST_CHECK_EQUAL(xData_arb_binary.searchLocal(xyPosition), size_t(1));
+  BOOST_CHECK_EQUAL(xData_arb_binary.searchLocal(xyPosition), size_t(0));
```
Comment (asalzburger) [inline Tests/UnitTests/Core/Utilities/BinningDataTests.cpp:200]: This was a buggy test covering a buggy result. `xData.bin(xyPosition)` has to result in bin 0, as it is `| 0 - 1|` and the the x value is 0.5,

Comment (paulgessinger) [reply Tests/UnitTests/Core/Utilities/BinningDataTests.cpp:200]: Ah, nice

**Review by paulgessinger — APPROVED** [review]
Seems fine to me.

**Comment by paulgessinger** [comment]
This changes the public API, but we probably wouldn't consider this breaking, right @asalzburger ?

**Pull request merged**